File: src/main/c/executeQuery.c
// Tested with
//   Apple clang version 12.0.0 (clang-1200.0.32.2)
//   Target: x86_64-apple-darwin19.6.0
// Compile from the project roots with
//   gcc -Wall -Ltarget -Itarget target/libneo4j.dylib src/main/c/executeQuery.c -o target/executeQuery
// And run as
//   target/executeQuery

#include <stdio.h>
#include <stdlib.h>
#include "libneo4j.h"

int main(void) {
	
	int ret;
	graal_create_isolate_params_t isolate_params;
	graal_isolate_t* isolate;
	graal_isolatethread_t *thread = NULL;

	ret = graal_create_isolate(&isolate_params, &isolate, &thread);
	if( ret != 0) {
		fprintf(stderr, "graal_create_isolate: %d\n", ret);
		exit(0);
	}
	
	int count = execute_query(thread, "bolt://localhost:7687", "secret", "MATCH (m:Movie) RETURN m");
	fprintf(stdout, "Number of movies printed: %d\n", count);
	
	if (graal_detach_thread(thread) != 0) {
	  fprintf(stderr, "graal_detach_thread error\n");
	  return 1;
	}
}

File: src/main/ruby/executeQuery.rb
# Tested with
# ruby --version
# ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-darwin19]
#
# gem install ffi
# ruby src/main/ruby/executeQuery.rb

require 'ffi'

class GRAAL_CREATE_ISOLATE_PARAMS_T < FFI::Struct

  layout :version,:int,
     :reserved_address_space_size,:long,
     :auxiliary_image_path,:string,
     :auxiliary_image_reserved_space_size,:long
end

class GRAAL_ISOLATE_T < FFI::Struct
end

class GRAAL_ISOLATETHREAD_T < FFI::Struct
end

module LibNeo4j extend FFI::Library

  ffi_lib '.target/libneo4j.so' if RUBY_PLATFORM =~ /linux/
  ffi_lib './target/libneo4j.dylib' if RUBY_PLATFORM =~ /darwin/

  attach_function :graal_create_isolate, [ GRAAL_CREATE_ISOLATE_PARAMS_T.by_ref, :pointer, :pointer], :int
  attach_function :graal_detach_thread, [ :pointer ], :int  
  attach_function :ffi_execute_query, :execute_query, [:pointer, :string, :string, :string ], :long
    
  class << self
    def execute_query(uri, password, query)

      isolate_params = GRAAL_CREATE_ISOLATE_PARAMS_T.new

      cnt = 0
      FFI::MemoryPointer.new(:pointer) do |isolate|
        FFI::MemoryPointer.new(:pointer) do |thread|
          graal_create_isolate(isolate_params, isolate, thread)
          cnt = ffi_execute_query(thread.get_pointer(0), uri, password, query)
          graal_detach_thread(thread.get_pointer(0))
        end
      end
      cnt
    end

    # Make the attached method private and only use our orchestrated version
    private :ffi_execute_query, :graal_create_isolate, :graal_detach_thread
  end
end

LibNeo4j.execute_query('bolt://localhost:7687', 'secret', 'MATCH (m:Movie) RETURN m')
